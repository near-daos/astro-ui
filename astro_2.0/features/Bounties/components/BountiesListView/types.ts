import { Bounty } from 'types/bounties';
import { ReactNode } from 'react';
import { ProposalVariant } from 'types/proposal';

export type SectionItem = {
  id: string;
  title: string;
  proposer: string;
  proposalId: string;
  bounty?: Bounty;
  link: {
    pathname: string;
    query: {
      dao: string;
      bounty?: string;
      proposal?: string;
    };
  };
  completeHandler?: (
    id: number,
    variant: ProposalVariant.ProposeDoneBounty
  ) => void;
  content: ReactNode;
};
