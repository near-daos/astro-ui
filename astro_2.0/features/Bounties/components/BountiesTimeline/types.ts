export type TimelineMilestoneType =
  | 'Proposal Created'
  | 'Bounty Created'
  | 'Claim'
  | 'Pending Approval'
  | 'Complete Claim'
  | 'Claim Deadline'
  | 'Complete Bounty';

export type TimelineMilestone = {
  type: TimelineMilestoneType;
  date: Date;
  tooltip: string;
  color?: string;
};

export type TimelineClaim = {
  id: string;
  milestones: TimelineMilestone[];
  title: string;
  color: string;
  minDate?: Date;
  maxDate?: Date;
};

export type TimelineGroup = {
  id: string;
  name: string;
  isOpen: boolean;
  milestones: TimelineMilestone[];
  claims: TimelineClaim[];
  minDate?: Date;
  maxDate?: Date;
};

export type TimelineGranularity = 'month' | 'day' | 'hour';
