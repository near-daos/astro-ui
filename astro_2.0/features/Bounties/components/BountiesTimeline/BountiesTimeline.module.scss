@import 'styles/mixins';
@import 'styles/typography';

.root {
  background: var(--color-white);
  border-radius: 8px;
  box-shadow: 3px 2px 24px var(--color-black-opacity-1);
  display: flex;
  flex-direction: column;
  overflow: hidden;
  position: relative;
}

.timelineControls {
  align-items: center;
  display: flex;
  justify-content: space-between;
}

.groups {
  background-color: var(--color-white);
  border-right: 1px solid var(--color-neutral-40);
  bottom: 0;
  left: 0;
  position: absolute;
  top: 0;
  width: 106px;
  z-index: 100;
}

.groupsHeader {
  @extend %caption2;
  border-bottom: 1px solid var(--color-neutral-40);
  height: 55px;
  padding: 20px 16px;
}

.group {
  @extend %caption2;
  align-items: center;
  cursor: pointer;
  display: flex;
  height: 26px;
  margin: 8px 0;
  padding: 0 4px;

  &::first-letter {
    text-transform: capitalize;
  }
}

.groupIcon {
  margin: 0 2px;
  width: 12px;
}

.groupName {
  @include ellipse-text;
}

.claimName {
  @extend %caption2;
  color: var(--color-neutral-60);
  cursor: default;
  font-size: 10px;
  padding: 0 4px;
}

.content {
  display: flex;
  height: 100%;
  margin: 0 0 0 106px;
  -ms-overflow-style: none;
  overflow-x: auto;
  overflow-y: hidden;
  scrollbar-width: none;
}

.content::-webkit-scrollbar {
  display: none;
}

.topColumn {
  text-align: center;
}

.topColumnLabel {
  @extend %caption2;
  padding: 8px 0;
}

.subColumns,
.dataColumns {
  display: flex;
}

.subColumns {
  border-bottom: 1px solid var(--color-neutral-40);
}

.dataColumns {
  margin: 8px 0;
}

.column {
  @extend %caption2;
  @include centralize;
  color: var(--color-neutral-60);
  height: 26px;
  min-width: 32px;
  padding: 4px;
  white-space: nowrap;
  width: 32px;
}

.subColumn {
  border-right: 1px solid var(--color-neutral-20);
  border-top: 1px solid var(--color-neutral-20);
  height: 20px;
  position: relative;
  z-index: 0;
}

.subColumn::after {
  border-right: 1px solid var(--color-neutral-20);
  content: '';
  position: absolute;
  right: -1px;
  top: 0;
  width: 32px;
  z-index: -1;
}

.dataColumn {
  position: relative;
}

.dataColumn::after {
  border-right: 1px solid var(--color-neutral-20);
  content: '';
  height: 42px;
  position: absolute;
  right: 0;
  top: -8px;
  width: 32px;
  z-index: 10;
}

.showTrack::before {
  background-color: currentColor;
  content: '';
  height: 1px;
  position: absolute;
  transform: translateY(-0.5px);
  width: 100%; // calc(100% + 32px);
  z-index: 20;
}

.showIncompleteTrack::before {
  border-bottom: 1px dashed currentColor;
  content: '';
  height: 1px;
  position: absolute;
  transform: translateY(-0.5px);
  width: 29px;
  z-index: 20;
}

.lastHeaderColumn::after {
  background-color: var(--color-primary-10);
  border-right: 1px solid var(--color-neutral-40);
  top: -36px;
  width: 33px;
}

.lastColumn::after {
  background-color: var(--color-primary-10);
  border-right: 1px solid var(--color-neutral-40);
}

.firstMilestone::before {
  left: 50%;
  right: unset;
  width: 50%;
}

.lastMilestone::before {
  left: unset;
  right: 50%;
  width: 50%;
}

.singleMilestone::before {
  width: 0;
}
