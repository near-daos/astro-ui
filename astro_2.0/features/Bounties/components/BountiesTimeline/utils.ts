function hslToHex(hue: number, sat: number, lig: number) {
  const l = lig / 100;

  const a = (sat * Math.min(l, 1 - l)) / 100;
  const f = (n: number) => {
    const k = (n + hue / 30) % 12;
    const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);

    return Math.round(255 * color)
      .toString(16)
      .padStart(2, '0');
  };

  return `#${f(0)}${f(8)}${f(4)}`;
}

export function generateHslaColors(
  saturation: number,
  lightness: number,
  alpha: number,
  amount: number
): string[] {
  const colors: string[] = [];
  const huedelta = Math.trunc(360 / amount);

  for (let i = 0; i < amount; i += 1) {
    const hue = i * huedelta;

    colors.push(hslToHex(hue, saturation, lightness));
  }

  return colors;
}
