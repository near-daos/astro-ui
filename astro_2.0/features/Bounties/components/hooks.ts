import { useCallback, useRef, useState } from 'react';
import { useRouter } from 'next/router';
import { useAsyncFn, useLocalStorage, useLocation } from 'react-use';
import axios, { CancelTokenSource } from 'axios';

import { SputnikHttpService } from 'services/sputnik';
import { useModal } from 'components/modal';
import { ConfirmActionModal } from 'astro_2.0/components/ConfirmActionModal';

import { Bounty, BountyContext, BountyProposal } from 'types/bounties';
import { VoteAction } from 'types/proposal';

import { NOTIFICATION_TYPES, showNotification } from 'features/notifications';

import { useWalletContext } from 'context/WalletContext';
import { PaginationResponse } from 'types/api';
import { VOTE_ACTION_SOURCE_PAGE } from 'constants/votingConstants';

export function useBountyControls(
  daoId: string,
  bounty?: Bounty
): {
  handleClaim: () => void;
  handleUnclaim: () => void;
  loading: boolean;
} {
  const router = useRouter();
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();
  const [loading, setLoading] = useState(false);

  const [showModal] = useModal(ConfirmActionModal);

  const onSuccessHandler = useCallback(async () => {
    router.reload();
  }, [router]);

  const handleClaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Claim',
    });

    if (res?.length && bounty) {
      setLoading(true);

      const dao = await SputnikHttpService.getDaoById(daoId);

      if (!dao) {
        return;
      }

      try {
        setVoteActionSource(pathname);

        await nearService?.claimBounty(daoId, {
          bountyId: bounty?.bountyId,
          deadline: bounty?.maxDeadline,
          bountyBond: dao.policy.bountyBond,
          gas: res[0],
          tokenId: bounty?.token,
        });

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    daoId,
    setVoteActionSource,
    pathname,
    nearService,
    onSuccessHandler,
  ]);

  const handleUnclaim = useCallback(async () => {
    const res = await showModal({
      title: 'Confirm Your Unclaim',
    });

    if (res?.length && bounty) {
      try {
        setLoading(true);
        setVoteActionSource(pathname);
        await nearService?.unclaimBounty(daoId, bounty?.bountyId);

        await onSuccessHandler();
      } catch (err) {
        setLoading(false);
        showNotification({
          type: NOTIFICATION_TYPES.ERROR,
          lifetime: 5000,
          description: err.message,
        });
      }
    }
  }, [
    showModal,
    bounty,
    setVoteActionSource,
    pathname,
    nearService,
    daoId,
    onSuccessHandler,
  ]);

  return {
    handleClaim,
    handleUnclaim,
    loading,
  };
}

export function useBountyVoting(
  daoId: string,
  proposal: BountyProposal
): {
  handleVote: (vote: VoteAction) => void;
  loading: boolean;
} {
  const router = useRouter();
  const [showModal] = useModal(ConfirmActionModal);
  const { nearService } = useWalletContext();
  const [, setVoteActionSource] = useLocalStorage(VOTE_ACTION_SOURCE_PAGE);
  const { pathname } = useLocation();

  const [{ loading }, handleVote] = useAsyncFn(
    async (vote: VoteAction) => {
      const res = await showModal({
        title: 'Confirm Your Vote',
      });

      if (res?.length) {
        setVoteActionSource(pathname);
        await nearService?.vote(daoId, proposal.proposalId, vote, res[0]);
        await router.reload();
      }
    },
    [daoId, proposal, router, nearService, pathname, setVoteActionSource]
  );

  return {
    handleVote,
    loading,
  };
}

export function useBountySearch(): {
  handleSearch: (
    val: string
  ) => Promise<PaginationResponse<BountyContext[]> | null>;
  loading: boolean;
} {
  const router = useRouter();
  const daoId = router.query.dao as string;
  const { accountId } = useWalletContext();
  const cancelTokenRef = useRef<CancelTokenSource | null>(null);

  const [{ loading }, handleSearch] = useAsyncFn(async query => {
    if (cancelTokenRef.current) {
      cancelTokenRef.current?.cancel('Cancelled by new req');
    }

    const { CancelToken } = axios;
    const source = CancelToken.source();

    cancelTokenRef.current = source;

    return SputnikHttpService.findBountyContext({
      daoId: daoId || '',
      accountId,
      query,
      cancelToken: source.token,
    });
  }, []);

  return {
    handleSearch,
    loading,
  };
}

export function useHideBounty(): {
  handleSubmit: () => void;
  handleSelect: (id: string) => void;
  selected: string[];
  loading: boolean;
} {
  const router = useRouter();
  const showHidden = router.query?.bountyFilter === 'hidden';
  const daoId = router.query.dao as string;
  const { accountId, nearService, pkAndSignature } = useWalletContext();
  const [selected, setSelected] = useState<string[]>([]);

  const [{ loading }, handleSubmit] = useAsyncFn(async () => {
    if (!pkAndSignature) {
      return;
    }

    try {
      const { publicKey, signature } = pkAndSignature;

      if (publicKey && signature) {
        await SputnikHttpService.toggleBountyContexts({
          accountId,
          publicKey,
          signature,
          daoId,
          ids: selected,
          isArchived: !showHidden,
        });

        await router.replace(router.asPath, undefined, {
          shallow: false,
        });

        setSelected([]);
      }
    } catch (err) {
      showNotification({
        type: NOTIFICATION_TYPES.ERROR,
        lifetime: 20000,
        description: err.message,
      });
    }
  }, [
    selected,
    daoId,
    showHidden,
    accountId,
    router,
    nearService,
    pkAndSignature,
  ]);

  const handleSelect = useCallback(
    (id: string) => {
      if (selected.includes(id)) {
        setSelected(selected.filter(item => item !== id));
      } else {
        setSelected([...selected, id]);
      }
    },
    [selected]
  );

  return {
    handleSubmit,
    handleSelect,
    selected,
    loading,
  };
}
