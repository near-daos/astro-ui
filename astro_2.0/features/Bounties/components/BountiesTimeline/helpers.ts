import { BountyClaim, BountyContext, BountyProposal } from 'types/bounties';
import {
  TimelineClaim,
  TimelineGranularity,
  TimelineGroup,
  TimelineMilestone,
} from 'astro_2.0/features/Bounties/components/BountiesTimeline/types';
import { DATA_SEPARATOR } from 'constants/common';
import {
  addDays,
  eachDayOfInterval,
  eachHourOfInterval,
  eachMinuteOfInterval,
  eachMonthOfInterval,
  endOfDay,
  endOfHour,
  endOfMonth,
  format,
  getDaysInMonth,
  isLastDayOfMonth,
  isSameDay,
  isSameHour,
  startOfDay,
  startOfHour,
  startOfMinute,
  startOfMonth,
} from 'date-fns';
import { toMillis } from 'utils/format';

const TOOLTIP_DATE_FORMAT = 'dd MMM, yyyy';

const COLORS = [
  '#DC88F5',
  '#89B1F9',
  '#51BAD1',
  '#EF7F31',
  '#E43AE3',
  '#57B99D',
  '#D32C1F',
  '#89B1F9',
  '#A87F3D',
  '#76140C',
  '#D65B26',
  '#B6BCC1',
  '#EC5281',
  '#EF87AA',
  '#925EB1',
  '#8060D9',
  '#0038DF',
  '#4581B2',
  '#46A8EE',
  '#65C978',
  '#697683',
];

function getColor(index: number) {
  return COLORS[index % COLORS.length];
}

function getClaimMilestones(
  claim: BountyClaim,
  color: string,
  proposal?: BountyProposal
) {
  const { accountId, startTime } = claim;
  const milestones: TimelineMilestone[] = [];
  const claimStartTime = toMillis(startTime);
  const claimStart = new Date(claimStartTime);

  if (claimStart) {
    milestones.push({
      type: 'Claim',
      date: claimStart,
      tooltip: `Claimed by ${accountId}`,
      color,
    });
  }

  if (proposal) {
    const proposalStatus = proposal.status;

    if (proposalStatus === 'InProgress') {
      milestones.push({
        type: 'Pending Approval',
        date: new Date(proposal.createdAt),
        tooltip: 'Pending Approval',
        color,
      });
    } else if (proposalStatus === 'Approved') {
      milestones.push({
        type: 'Complete Claim',
        date: new Date(proposal.createdAt),
        tooltip: `Claim completed by ${proposal.proposer}`,
        color,
      });
    }
    // todo - how cna we handle rejected/not approved claims?
  }

  if (claimStart && claim.deadline) {
    const hasApprovedProposal = !!milestones.find(
      prp => prp.type === 'Complete Claim'
    );

    if (!hasApprovedProposal) {
      const deadline = toMillis(claim.deadline);
      const claimEnd = new Date(claimStartTime + deadline);

      milestones.push({
        type: 'Claim Deadline',
        date: claimEnd,
        tooltip: `Deadline to complete ${accountId} claim ${format(
          claimEnd,
          TOOLTIP_DATE_FORMAT
        )}`,
        color,
      });
    }
  }

  return milestones;
}

function prepareClaims(context: BountyContext) {
  const claims: TimelineClaim[] = [];

  // Bounty claims
  if (context.bounty?.bountyClaims?.length > 0) {
    context.bounty.bountyClaims.forEach((claim, i) => {
      const { id } = claim;
      const proposal = context.bounty.bountyDoneProposals.find(
        _proposal => _proposal.bountyClaimId === id
      );

      const milestones = getClaimMilestones(claim, getColor(i), proposal);

      const sortedMilestones = milestones.slice().sort((a, b) => {
        if (a.date > b.date) {
          return 1;
        }

        if (a.date < b.date) {
          return -1;
        }

        return 0;
      });

      const minDate = sortedMilestones[0]?.date;
      const maxDate = sortedMilestones[sortedMilestones.length - 1]?.date;

      claims.push({
        id: claim.id,
        milestones,
        title: claim.accountId,
        color: getColor(i),
        minDate,
        maxDate,
      });
    });
  }

  return claims;
}

function accumulateClaimsMilestones(
  milestones: TimelineMilestone[],
  claims: TimelineClaim[]
) {
  return claims.reduce(
    (res, item) => {
      res.push(...item.milestones);

      return res;
    },
    [...milestones]
  );
}

export function prepareTimelineDataset(data: BountyContext[]): TimelineGroup[] {
  return data.map(context => {
    const item = context.bounty || context.proposal;
    let milestones: TimelineMilestone[] = [];
    const [description] = item.description.split(DATA_SEPARATOR);

    if (context.proposal) {
      // Create proposal milestone
      const proposalCreated = new Date(context.proposal.createdAt);

      milestones.push({
        type: 'Proposal Created',
        date: proposalCreated,
        tooltip: `Proposal Created ${format(
          proposalCreated,
          TOOLTIP_DATE_FORMAT
        )}`,
      });
    }

    if (context.bounty) {
      // Create bounty milestone
      const bountyCreated = new Date(context.bounty.createdAt);

      milestones.push({
        type: 'Bounty Created',
        date: bountyCreated,
        tooltip: `Bounty Created ${format(bountyCreated, TOOLTIP_DATE_FORMAT)}`,
      });
    }

    if (Number(context.bounty?.times) === 0) {
      // Complete bounty milestone
      let latestDate: Date | null = null;

      context.bounty.bountyDoneProposals
        .filter(_proposal => _proposal.status === 'Approved')
        .forEach(_proposal => {
          const updatedAt = new Date(_proposal.updatedAt);

          if (!latestDate || latestDate < updatedAt) {
            latestDate = updatedAt;
          }
        });

      if (latestDate) {
        milestones.push({
          type: 'Complete Bounty',
          date: latestDate,
          tooltip: `Bounty Completed ${format(
            latestDate,
            TOOLTIP_DATE_FORMAT
          )}`,
        });
      }
    }

    const claims = prepareClaims(context);

    milestones = accumulateClaimsMilestones(milestones, claims);

    const sortedMilestones = milestones.slice().sort((a, b) => {
      if (a.date > b.date) {
        return 1;
      }

      if (a.date < b.date) {
        return -1;
      }

      return 0;
    });

    const groupMinDate = sortedMilestones[0]?.date;
    const groupMaxDate = sortedMilestones[sortedMilestones.length - 1]?.date;

    return {
      id: context.id,
      isOpen: false,
      name: description || '',
      milestones,
      claims,
      minDate: groupMinDate,
      maxDate: groupMaxDate,
    };
  });
}

export function getTimelineRange(dataset: TimelineGroup[]): [Date, Date] {
  let min: Date | null = null; // new Date();
  let max: Date | null = null; // addMonths(min, 1);

  dataset.forEach(group => {
    group.milestones.forEach(milestone => {
      if (!min || min > milestone.date) {
        min = milestone.date;
      }

      if (!max || max < milestone.date) {
        max = milestone.date;
      }
    });
  });

  if (!min) {
    min = new Date();
  }

  if (!max) {
    max = addDays(min, 2);
  }

  return [min, max];
}

export function getTopColumns(
  start: Date,
  end: Date,
  granularity: TimelineGranularity
): Date[] {
  switch (granularity) {
    case 'hour': {
      return eachHourOfInterval({ start, end });
    }
    case 'day': {
      return eachDayOfInterval({ start, end });
    }
    default: {
      return eachMonthOfInterval({ start, end });
    }
  }
}

export function getColumns(
  start: Date,
  end: Date,
  granularity: TimelineGranularity
): Date[] {
  switch (granularity) {
    case 'hour': {
      return eachMinuteOfInterval({ start, end });
    }
    case 'day': {
      return eachHourOfInterval({ start, end });
    }
    default: {
      return eachDayOfInterval({ start, end });
    }
  }
}

export function formatColumnLabel(
  date: Date,
  granularity: TimelineGranularity
): string {
  switch (granularity) {
    case 'hour': {
      return format(date, 'mm');
    }
    case 'day': {
      return format(date, 'H');
    }
    default: {
      return format(date, 'd');
    }
  }
}

export function getTopColumnsWidth(
  date: Date,
  granularity: TimelineGranularity
): number {
  switch (granularity) {
    case 'hour': {
      return 60 * 32;
    }
    case 'day': {
      return 24 * 32;
    }
    default: {
      return getDaysInMonth(date) * 32;
    }
  }
}

export function getRangeColumns(
  date: Date,
  granularity: TimelineGranularity
): Date[] {
  switch (granularity) {
    case 'hour': {
      return eachMinuteOfInterval({
        start: startOfHour(date),
        end: endOfHour(date),
      });
    }
    case 'day': {
      return eachHourOfInterval({
        start: startOfDay(date),
        end: endOfDay(date),
      });
    }
    default: {
      return eachDayOfInterval({
        start: startOfMonth(date),
        end: endOfMonth(date),
      });
    }
  }
}

export function getMilestonesForDate(
  milestones: TimelineMilestone[],
  date: Date,
  granularity: TimelineGranularity
): TimelineMilestone[] {
  let compareFn = isSameDay;

  switch (granularity) {
    case 'day': {
      compareFn = isSameHour;

      break;
    }
    default: {
      compareFn = isSameDay;

      break;
    }
  }

  return (
    milestones?.filter(item => {
      return compareFn(date, item.date);
    }) ?? []
  );
}

export function getGroupDateStart(
  date: Date | undefined,
  granularity: TimelineGranularity
): Date | null {
  if (!date) {
    return null;
  }

  switch (granularity) {
    case 'hour': {
      return startOfMinute(date);
    }
    case 'day': {
      return startOfHour(date);
    }
    default: {
      return startOfDay(date);
    }
  }
}

export function isEndOfGranularityPeriod(
  date: Date,
  granularity: TimelineGranularity
): boolean {
  switch (granularity) {
    case 'hour': {
      return date.getMinutes() === 59;
    }
    case 'day': {
      return date.getHours() === 23;
    }
    default: {
      return isLastDayOfMonth(date);
    }
  }
}
