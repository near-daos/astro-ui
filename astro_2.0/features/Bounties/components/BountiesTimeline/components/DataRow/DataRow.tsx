import React, { FC } from 'react';
import cn from 'classnames';

import {
  getMilestonesForDate,
  isEndOfGranularityPeriod,
} from 'astro_2.0/features/Bounties/components/BountiesTimeline/helpers';
import { Milestone } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/Milestone';
import { StackedMilestones } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/StackedMilestones';
import {
  TimelineGranularity,
  TimelineMilestone,
} from 'astro_2.0/features/Bounties/components/BountiesTimeline/types';

import styles from 'astro_2.0/features/Bounties/components/BountiesTimeline/BountiesTimeline.module.scss';

interface DataRowProps {
  data: TimelineMilestone[];
  rangeColumns: Date[];
  granularity: TimelineGranularity;
  minDate: Date | null;
  maxDate: Date | null;
  color?: string;
  isGroupRow: boolean;
}

export const DataRow: FC<DataRowProps> = React.memo(
  ({
    data,
    rangeColumns,
    granularity,
    minDate,
    maxDate,
    color,
    isGroupRow,
  }) => {
    // const deadlineClaimMilestone = data.find(
    //   item => item.type === 'Claim Deadline'
    // );
    const approveClaimMilestone = data.find(
      item => item.type === 'Pending Approval'
    );
    const firstMilestone = data[0];
    const lastMilestone = data[data.length - 1];

    return (
      <div className={styles.dataColumns}>
        {rangeColumns.map((columnDate, i) => {
          const milestones = getMilestonesForDate(
            data,
            columnDate,
            granularity
          );
          const isEndOfPeriod = isEndOfGranularityPeriod(
            columnDate,
            granularity
          );

          const showTrack =
            minDate &&
            maxDate &&
            columnDate >= minDate &&
            columnDate <= maxDate;
          const showIncompleteTrack =
            !isGroupRow &&
            showTrack &&
            (!approveClaimMilestone || columnDate > approveClaimMilestone.date);

          const isFirstMilestone =
            firstMilestone &&
            showTrack &&
            milestones.length &&
            milestones[0].date === firstMilestone.date;
          const isLastMilestone =
            showTrack &&
            lastMilestone &&
            milestones.length &&
            milestones[milestones.length - 1].date === lastMilestone.date;

          return (
            <div
              /* eslint-disable-next-line react/no-array-index-key */
              key={i}
              data-milestone={milestones?.length > 0}
              style={{ color }}
              className={cn(styles.column, styles.dataColumn, {
                [styles.showTrack]: showTrack && !showIncompleteTrack,
                [styles.showIncompleteTrack]: showIncompleteTrack,
                [styles.lastColumn]: isEndOfPeriod,
                [styles.firstMilestone]: isFirstMilestone,
                [styles.lastMilestone]: isLastMilestone,
                [styles.singleMilestone]: isLastMilestone && isFirstMilestone,
              })}
            >
              {milestones && milestones.length === 1 && (
                <Milestone data={milestones[0]} color={milestones[0].color} />
              )}
              {milestones && milestones.length > 1 && (
                <StackedMilestones data={milestones} />
              )}
            </div>
          );
        })}
      </div>
    );
  }
);
