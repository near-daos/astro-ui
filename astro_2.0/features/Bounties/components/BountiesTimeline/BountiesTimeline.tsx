import React, {
  FC,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';

import { BountyContext } from 'types/bounties';
import {
  getGroupDateStart,
  getRangeColumns,
  getTimelineRange,
  getTopColumns,
  prepareTimelineDataset,
} from 'astro_2.0/features/Bounties/components/BountiesTimeline/helpers';
import { DataRow } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/DataRow';
import { TimelineLegend } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/TimelineLegend';
import { TimelineRangeToggle } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/TimelineRangeToggle';
import {
  TimelineGranularity,
  TimelineGroup,
} from 'astro_2.0/features/Bounties/components/BountiesTimeline/types';
import { TimelineGroups } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/TimelineGroups';
import { TimelineHeader } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/TimelineHeader';
import { NoResultsView } from 'astro_2.0/components/NoResultsView';

import useQuery from 'hooks/useQuery';

import styles from './BountiesTimeline.module.scss';

interface BountiesTimelineProps {
  data: BountyContext[];
}

export const BountiesTimeline: FC<BountiesTimelineProps> = ({ data }) => {
  const contentRef = useRef<HTMLDivElement>(null);
  const { query } = useQuery<{
    bountyFilter: string;
    bountySort: string;
  }>({ shallow: false });
  const [dataset, setDataset] = useState<TimelineGroup[]>([]);
  const [granularity, setGranularity] = useState<TimelineGranularity>('month');
  const [min, max] = useMemo(() => getTimelineRange(dataset), [dataset]);
  const topColumns = useMemo(
    () => getTopColumns(min, max, granularity),
    [granularity, max, min]
  );

  const toggleGroup = useCallback(
    groupId => {
      const updatedDataSet = dataset.map(group => {
        if (group.id === groupId) {
          return {
            ...group,
            isOpen: !group.isOpen,
          };
        }

        return group;
      });

      setDataset(updatedDataSet);
    },
    [dataset]
  );

  const toggleRange = useCallback(val => {
    setGranularity(val);
  }, []);

  useEffect(() => {
    // We want milestone to be visible on load
    if (contentRef.current) {
      const container = contentRef.current;
      const milestone = document.querySelector(
        '[data-milestone]'
      ) as HTMLDivElement;

      if (milestone && container) {
        container.scrollLeft = milestone.offsetLeft + 64;
      }
    }
  }, []);

  useEffect(() => {
    setDataset(prepareTimelineDataset(data));
  }, [data, query]);

  if (!dataset.length) {
    return <NoResultsView title="No results found" />;
  }

  return (
    <>
      <div className={styles.timelineControls}>
        <TimelineLegend />
        <TimelineRangeToggle onChange={toggleRange} selected={granularity} />
      </div>
      <div className={styles.root}>
        <div className={styles.content} ref={contentRef}>
          <TimelineGroups dataset={dataset} toggleGroup={toggleGroup} />
          {topColumns.map((item, i) => {
            const rangeColumns = getRangeColumns(item, granularity);

            return (
              <div
                /* eslint-disable-next-line react/no-array-index-key */
                key={i}
                className={styles.topColumn}
              >
                <TimelineHeader
                  item={item}
                  rangeColumns={rangeColumns}
                  granularity={granularity}
                />
                {dataset.map(group => {
                  const minDate = getGroupDateStart(group.minDate, granularity);
                  const maxDate = getGroupDateStart(group.maxDate, granularity);

                  return (
                    <React.Fragment key={group.id}>
                      <DataRow
                        data={group.milestones}
                        isGroupRow
                        rangeColumns={rangeColumns}
                        granularity={granularity}
                        minDate={minDate}
                        maxDate={maxDate}
                      />
                      {group.isOpen &&
                        group.claims.length > 0 &&
                        group.claims.map(claim => {
                          const claimMinDate = getGroupDateStart(
                            claim.minDate,
                            granularity
                          );
                          const claimMaxDate = getGroupDateStart(
                            claim.maxDate,
                            granularity
                          );

                          return (
                            <DataRow
                              key={claim.id}
                              isGroupRow={false}
                              color={claim.color}
                              data={claim.milestones}
                              rangeColumns={rangeColumns}
                              granularity={granularity}
                              minDate={claimMinDate}
                              maxDate={claimMaxDate}
                            />
                          );
                        })}
                    </React.Fragment>
                  );
                })}
              </div>
            );
          })}
        </div>
      </div>
    </>
  );
};
