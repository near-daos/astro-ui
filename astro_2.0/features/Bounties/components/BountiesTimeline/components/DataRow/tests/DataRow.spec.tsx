import { render } from 'jest/testUtils';

import { TimelineMilestone } from 'astro_2.0/features/Bounties/components/BountiesTimeline/types';

import { DataRow } from 'astro_2.0/features/Bounties/components/BountiesTimeline/components/DataRow';

describe('DataRow', () => {
  it('Should render component', () => {
    const item = {
      type: 'Pending Approval',
    } as unknown as TimelineMilestone;

    const { container } = render(
      <DataRow
        data={[item]}
        rangeColumns={[new Date(2222, 12, 12)]}
        granularity="day"
        minDate={null}
        maxDate={null}
        isGroupRow={false}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
