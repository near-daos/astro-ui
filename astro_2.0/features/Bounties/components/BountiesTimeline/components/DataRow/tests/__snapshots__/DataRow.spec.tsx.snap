// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataRow Should render component 1`] = `
<div>
  <div
    class="dataColumns"
  >
    <div
      class="column dataColumn"
      data-milestone="false"
    />
  </div>
</div>
`;
